### PR TITLE
[Python] Initial support of cloud API

### PR DESCRIPTION
**Pull request by eddyxu**
Support connect with remote database, and implement Search API

### CODE REVIEW

File: python/lancedb/remote/client.py [hunk 1]
```diff
@@ -35,29 +35,32 @@ def wrapped(self, *args, **kwargs):
 
 @attr.define(slots=False)
 class RestfulLanceDBClient:
-    url: str
+    db_name: str
+    region: str
+    api_key: str
```
Comment (chebbyChefNEQ) [inline python/lancedb/remote/client.py:40]: we should make sure this field doesn't get accidentally printed somewhere.

something like 
```
class Password(str):
    def __repr__(self):
        return "****"

   def __str__(self):
        return "****"
```

Comment (eddyxu) [reply python/lancedb/remote/client.py:40]: can i just do __repr__ and __str__ on this Query object?

File: python/lancedb/remote/db.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+        name: str
+            The name of the table.
+
+        Returns
+        -------
+        A LanceTable object representing the table.
+        """
+        from .table import RemoteTable
+
+        return RemoteTable(self, name)
```
Comment (chebbyChefNEQ) [inline python/lancedb/remote/db.py:57]: let's add a todo to check for existence?

Comment (eddyxu) [reply python/lancedb/remote/db.py:59]: Done.

File: python/lancedb/__init__.py [hunk 3]
```diff
@@ -11,16 +11,24 @@
...
+from typing import Optional
 
+from .db import URI, DBConnection, LanceDBConnection
+from .remote.db import RemoteDBConnection
 
-def connect(uri: URI) -> LanceDBConnection:
-    """Connect to a LanceDB instance at the given URI
+
+def connect(
+    uri: URI, api_key: Optional[str] = None, region: str = "us-west-2"
```
Comment (chebbyChefNEQ) [inline python/lancedb/__init__.py:21]: Nit: let's add a * after uri requiring args to be named

File: python/lancedb/remote/client.py [hunk 4]
```diff
@@ -35,29 +36,32 @@ def wrapped(self, *args, **kwargs):
...
     def session(self) -> aiohttp.ClientSession:
-        parsed = urllib.parse.urlparse(self.url)
-        scheme = parsed.scheme
-        if not scheme.startswith("lancedb"):
-            raise ValueError(
-                f"Invalid scheme: {scheme}, must be like lancedb+<flavor>://"
-            )
-        flavor = scheme.split("+")[1]
-        url = f"{flavor}://{parsed.hostname}:{parsed.port}"
+        url = f"https://{self.db_name}.{self.region}.api.lancedb.com"
```
Comment (chebbyChefNEQ) [inline python/lancedb/remote/client.py:46]: Nit: add a plaintext argument where we use http over https

**Review by chebbyChefNEQ — APPROVED** [review]
Just two nits

File: python/lancedb/__init__.py (hunk 3 above)
Comment (eddyxu) [reply python/lancedb/__init__.py:21]: Fixed.

**Pull request merged**